### DISCUSSION TITLE
[이동규] Best Practice

### DISCUSSION THREAD

**Discussion by tenenger7125** [pre-onboarding-frontend-11th-team-9/pre-onboarding-11th-1-9#4] [Best Practice]
# 목차

- [목차](#목차)
- [Best Practice 선정 기준](#best-practice-선정-기준)
- [동료 Repository](#동료-repository)
- [코드 선정 기준](#코드-선정-기준)
- [Assignment1 : validation](#assignment1--validation)
- [Assignment2 : auth \& redirect](#assignment2--auth--redirect)
- [Assignment3 : redirect](#assignment3--redirect)
- [Assignment4 : todo list](#assignment4--todo-list)
- [Assignment5 : todo list (modify)](#assignment5--todo-list-modify)
- [기타](#기타)
- [궁금한 점](#궁금한-점)

# Best Practice 선정 기준

- 재사용과 가독성향상을 위해 커스텀 훅 패턴을 선호합니다.
- 유지보수와 가독성을 중점으로 확인합니다.
- 디자인은 주관적으로 평가합니다.

# 동료 Repository

- [김나현](https://github.com/reezekim/wanted-pre-onboarding-frontend) : JavaScript, CSS Module
- [김보라](https://github.com/rockbell89/wanted-pre-onboarding-frontend) : TypeScript, CSS(SCSS) Module
- [박용준](https://github.com/yoopark/wanted-pre-onboarding-frontend/tree/develop) : TypeScript, emotion
- [이동규(본인)](https://github.com/tenenger7125/wanted-pre-onboarding-frontend) : JavaScript, styled-components
- [조은선](https://github.com/es39/wanted-pre-onboarding-frontend) : JavaScript, styled-components
- [편경진](https://github.com/kjindev/wanted-pre-onboarding-frontend) : JavaScript, tailwind

# 코드 선정 기준

- 디스코드에 댓글로 올라온 순서대로 Repository 확인했고, 중복된 내용의 코드의 경우 첫번째로 확인한 동료의 코드만 인용했습니다.
- 내 코드가 좋다고 판단되고 다른 사람이 나와 유사한 코드라면 다른 사람의 코드를 선정했습니다.
- 조은선 ➡️ 김나현 ➡️ 이동규(본인) ➡️ 편경진 ➡️ 박용준 ➡️ 김보라

# Assignment1 : validation

> - 이메일과 비밀번호의 유효성 검사기능을 구현해주세요
>   - 이메일 조건: `@` 포함
>   - 비밀번호 조건: 8자 이상
>   - 입력된 이메일과 비밀번호가 위 조건을 만족할 때만 버튼이 활성화 되도록 해주세요
>   - ~~보안 상 실제 사용하고 계신 이메일과 패스워드말고 테스트용 이메일, 패스워드 사용을 권장드립니다.~~

- **Best Practice Code : 이동규(본인)**
- 로그인/회원가입 폼에서 공동으로 사용된다는 점에서, 재사용을 위해 유효성 검사 유틸 함수를 만드는 것이 좋다고 생각합니다.
- 추가로, 응집도를 위해 객체의 메서드로 표현하는 것이 좋다고 생각합니다.

```js
const validate = {
  email(email) {
    return email.includes("@");
  },
  password(password) {
    return password.length >= 8;
  },
};
```

- change 이벤트 핸들러 동작시 유효성 검사를 하는데, 과도한 리렌더링이 발생해서 `lodash`의 `debounce`를 활용하여 리렌더링을 최소화하는게 좋다고 생각합니다.
- 다른 분들은 이벤트 핸들러를 `useCallback`나 `useMemo`으로 기억하여 리렌더링시 함수를 재정의하는 것을 막는 경우가 있었는데, 기억하는 자원의 소모가 더 큰 것 같아서 사용하지 않는게 좋다고 생각합니다. `sort`와 같이 데이터 정렬할 때는 기억하는 것이 필요할 것 같습니다.

```js
const handleFieldChange = (e) => {
  const { name, value } = e.target;

  setForm({
    ...form,
    [name]: value,
  });
};
```

---

# Assignment2 : auth & redirect

> - 로그인 API를 호출하고, 올바른 응답을 받았을 때 `/todo` >경로로 이동해주세요

- **Best Practice Code : 박용준**
- routes가 여러 컴포넌트에서 사용된다는 점에서 상수로 관리하는게 좋다고 생각합니다.

```js
const ROUTES = {
  ROOT: "/",
  SIGN_UP: "/signup",
  SIGN_IN: "/signin",
  TODO: "/todo",
};
```

- try-catch, async-await 구문을 사용하여 동기적으로 동작하는 것 처럼 보이게만들어 가독성을 향상시킨 것이 좋다고 생각합니다.
- 고민할 점은, 에러핸들링을 위해 try-catch 구문을 컴포넌트안에 작성하니 가독성이 좋지 않는데, 더 좋은 방법이 있는지 고민해야할거 같습니다.

```js
const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    try {
      const res = await signin(formData);
      if (res.status === 200) {
        setAccessToken(res.data.access_token);
        navigate(ROUTES.TODO);
      }
    } catch (e: unknown) {
      if (isAxiosErrorFromWantedPreOnboardingServer<SigninResponse>(e)) {
        const { message } = e.response.data;
        if (message === 'Unauthorized') {
          alert('비밀번호가 틀렸습니다');
        } else {
          alert(message);
        }
      }
    }
}
```

---

- **Extra Practice Code : 편경진**
- useSign 커스텀 훅을 통해 로그인/회원가입 페이지 로직을 추상화하여 가독성을 향상시킨것은 좋다고 생각합니다.
- 고민할 점은, 로그인/회원가입 페이지 중 input field 값이 페이지마다 다른 경우에는 어떻게 처리해야할지 고민이 필요할 것 같습니다.

```js
function useSign() {
  const requestSign = async (event, type, inputEmail, inputPassword) => {
    event.preventDefault();
    try {
      const { status, data } = await axios({
        url: `https://www.pre-onboarding-selection-task.shop/auth/${type}`,
        method: "post",
        data: {
          email: inputEmail,
          password: inputPassword,
        },
      });
      if (type === "signin" && status === 200) {
        localStorage.setItem("token", data.access_token);
        window.location.replace("/todo");
      } else if (type === "signup" && status === 201) {
        window.location.replace("/signin");
      } else {
        alert("error");
      }
    } catch (err) {
      alert(err.response.data.message);
    }
  };

  return { requestSign };
}
```

---

> - 로그인 API는 로그인이 성공했을 시 Response Body에 JWT를 포함해서 응답합니다.

- **Best Practice Code : 김보라**
- 재사용측면에서 jwt 유틸 함수를 만드는 것이 좋다고 생각합니다.
- 추가로, 재사용측면에서 매개변수 없이 토큰을 얻을 수 있다는 점이 좋다고 생각합니다.
- 보완한다면, `TOKEN_NAME`변수를 함수 바깥에 선언하여 함수 호출마다 변수를 선언하는것을 막는것이 좋다고 생각합니다.

```js
const jwtStorageService = () => {
  const TOKEN_NAME = "access_token";

  const getToken = () => {
    return localStorage.getItem(TOKEN_NAME);
  };

  const setToken = (token: string) => {
    localStorage.setItem(TOKEN_NAME, token);
  };

  const removeToken = () => {
    localStorage.removeItem(TOKEN_NAME);
  };

  return {
    getToken,
    setToken,
    removeToken,
  };
};
```

---

> - 응답받은 JWT는 로컬 스토리지에 저장해주세요

- **Best Practice Code : 이동규(본인)**
- 재사용측면에서 로컬 스토리지 유틸 함수를 만드는 것이 좋다고 생각합니다.
- 추가로, 함수의 응집도를 위해 객체의 메서드로 갖는것이 더 좋다고 생각합니다.
- 궁금한점은, `getter`나 `setter`로 활용하는것이 나을지 궁금합니다.

```js
const localStorages = {
  setLocalStorage(key, value) {
    return localStorage.setItem(key, JSON.stringify(value));
  },
  getLocalStorage(key) {
    return JSON.parse(localStorage.getItem(key));
  },
};
```

---

# Assignment3 : redirect

> - 로그인 여부에 따른 리다이렉트 처리를 구현해주세요
>   - 로컬 스토리지에 토큰이 있는 상태로 `/` 페이지에 접속한다면 `/todo` 경로로 리다이렉트 시켜주세요
>   - 로컬 스토리지에 토큰이 없는 상태로 `/todo`페이지에 접속한다면 `/` 경로로 리다이렉트 시켜주세요

- **Best Practice Code : 이동규(본인)**
- 인증된 사람만 접근가능하거나 또는 인증된 사람만 접근불가능한 페이지를 구현하기 위해 Guard 컴포넌트로 분리한것이 좋다고 생각합니다.
- props로 경로를 받아 재사용 측면을 향상시키는 것이 좋다고 생각합니다.
- 보완할 점은, access_token을 인수를 넘겨주지 않고 추상화하여 access_token을 얻는 방법을 보완해야할것 같습니다.
  - [access_token 관련 jwt 로직](#best-practice)

```js
const PrivateRoute = ({ element, redirectTo }) => {
  const authenticated = getLocalStorage("access_token");

  return authenticated ? element : <Navigate to={redirectTo} />;
};
const PublicRoute = ({ element, redirectTo, restricted = false }) => {
  const authenticated = getLocalStorage("access_token");

  return authenticated && restricted ? <Navigate to={redirectTo} /> : element;
};
```

---

# Assignment4 : todo list

> - `/todo`경로에 접속하면 투두 리스트의 목록을 볼 수 있도록 해주세요
> - 리스트 페이지에는 투두 리스트의 내용과 완료 여부가 표시되어야 합니다.
> - 리스트 페이지에는 입력창과 추가 버튼이 있고, 추가 버튼을 누르면 입력창의 내용이 새로운 투두 리스트로 추가되도록 해주세요

- **Best Practice Code : 이동규(본인)**
- 추상화를 통해 가독성을 향상시킬 수 있어서 좋다고 생각합니다.
- 보완할 점은, useTodo가 여러번 호출되면 하나의 상태를 참조하지 않기 때문에 오류가 생긴다는 점, props drilling이 발생한다는 것을 고려해야할 것 같습니다.
  - useContext 또는 전역 상태관리 라이브러리 활용
- 고민할 점은, try-catch 구문을 이벤트 핸들러 내부에 작성하니 가독성이 좋지 않아서 대안을 고민해봐야할 것 같습니다.

```js
const useTodo = () => {
  const [todos, setTodos] = useState([]);

  const handleCreateTodo = async (e) => {
    e.preventDefault();

    try {
      const todo = await createTodo(e.target.newTodo.value.trim());

      e.target.newTodo.value = "";
      setTodos([...todos, todo]);
    } catch (err) {
      console.error(err);
    }
  };
  ...
}
```

# Assignment5 : todo list (modify)

> - 투두 리스트의 수정, 삭제 기능을 구현해주세요
>   - 투두 리스트의 개별 아이템 우측에 수정버튼이 존재하고 해당 버튼을 누르면 수정모드가 활성화되고 투두 리스트의 내용을 수정할 수 있도록 해주세요
>   - 수정모드에서는 개별 아이템의 우측에 제출버튼과 취소버튼이 표시되며 해당 버튼을 통해서 수정 내용을 제출하거나 수정을 취소할 수 있도록 해주세요
>   - 투두 리스트의 개별 아이템 우측에 삭제버튼이 존재하고 해당 버튼을 누르면 투두 리스트가 삭제되도록 해주세요

- **Best Practice Code : 이동규(본인)**
- 추상화를 통해 가독성을 향상시킬 수 있어서 좋다고 생각합니다.
- 보완할 점은, useTodo가 여러번 호출되면 하나의 상태를 참조하지 않기 때문에 오류가 생긴다는 점, props drilling이 발생한다는 것을 고려해야할 것 같습니다.
  - useContext 또는 전역 상태관리 라이브러리 활용
- 고민할 점은, try-catch 구문을 이벤트 핸들러 내부에 작성하니 가독성이 좋지 않아서 대안을 고민해봐야할 것 같습니다.

```js
const useTodo = () => {
  ...
  const handleUpdateTodo = async (id, todo, isCompleted) => {
    try {
      const updatedTodo = await updateTodo(id, todo, isCompleted);

      setTodos(todos.map((todo) => (todo.id === id ? { ...todo, ...updatedTodo } : todo)));
    } catch (err) {
      console.error(err);
    }
  };

  const handleDeleteTodo = async (id) => {
    try {
      await deleteTodo(id);

      setTodos(todos.filter((todo) => todo.id !== id));
    } catch (err) {
      console.error(err);
    }
  };

  useEffect(() => {
    (async () => {
      try {
        const todos = await getTodos();

        setTodos(todos);
      } catch (err) {
        console.error(err);
      }
    })();
  }, []);

  return { todos, handleCreateTodo, handleUpdateTodo, handleDeleteTodo };
};

export default useTodo;
```

---

# 기타

> - 스타일링

- **Best Practice Code : 조은선, 김나현, 편경진**
- 파일을 열때 스타일 컴포넌트가 하위에 위치하니 메인 컴포넌트를 한눈에 알아볼 수 있어 가독성이 향상된다고 생각합니다. -**조은선**-
- 고민하는 것, export default의 위치가 애매하게 중간에 위치하는 것과 최하단에 위치하는 것 중 어느 것이 좋은지 같이 고민해봐야할 거 같습니다.

```js
const SignIn = () => {...}
export default SignIn;

const Container = styled.main`
  display: flex;
  justify-content: center;
  align-items: center;
`;
```

- 스타일링 컴포넌트를 파일로 분리할 경우, 하나의 폴더 안에 동일한 코드를 참조하는 파일들을 넣는게 좋다고 생각합니다. -**김나현**-

- tailwind.css를 사용하니 컴포넌트 분리나 위치 걱정없어도 되니 깔끔해지게 만들어지는 것이 좋다고 생각합니다. -**편경진**-

---

> - UI/UX

- **Best Practice Code : 김나현**
- 로그인, 회원가입, todo 페이지의 UI/UX가 좋다고 생각합니다.

---

> - README.md 작성

- **Best Practice Code : 김나현, 조은선, 박용진**
- 코드를 보기 전, 폴더 구조를 설명하는 것이 좋은것 같다고 생각합니다. -**김나현**-
- 코드를 보기 전, 페이지 관련 설명이 좋은것 같다고 생각합니다. -**조은선**-
- 코드를 보기 전,무슨 기술스택 썼는지, 어떤것에 중점을 두고 개발했는지 설명하는 것이 좋은것 같다고 생각합니다. -**박용진**-

> - 초기 설정

- **Best Practice Code : 박용진**
- eslint, prettier를 이용한 코딩 컨벤션 유지, craco로 절대경로 추가하는게 좋다고 생각합니다.

# 궁금한 점

- 객체의 메서드를 활용하는 것이니 `jwtStorageService`가 함수가 아닌 객체로 정의하는 것은 어떨까요? -**김보라**-
- pages안 폴더안에 index.jsx로 만드신 이유가 무엇일까요? -**김보라**-
- provider 폴더안에는 어떤 것들이 추가로 들어갈 수 있나요? -**박용진**-
- isAxiosErrorFromWantedPreOnboardingServer는 어떻게 사용하는 건가요? -**박용진**-
- useLayoutEffect 쓴 이유가 무엇일까요? -**편경진**-

**Comment by rockbell89** [comment]
@tenenger7125 
- 객체의 메서드를 활용하는 것이니 jwtStorageService가 함수가 아닌 객체로 정의하는 것은 어떨까요? 
좋은 의견 감사합니다. 사실 저 패턴으로 작성한 장점이 없는 기능이긴 해서, 객체 리터럴 방식으로 정희하면 좋을 것 같습니다 
- pages안 폴더안에 index.jsx로 만드신 이유가 무엇일까요? 
이번 과제에서는 사용을 안하긴했지만,  상세페이지 혹은 추가 하위페이지 작업을 할때를 고려해서 적은 패턴이긴 합니다
```
/todo/index.tsx -  따로 이름을 정하지 않아도 기본 페이지
/todo/detail/index.tsx
```
페이지 뿐만 아니라 ts파일 컴포넌트 파일 작성할때도 이런 패턴을 사용하면, import 할때도 좀더 가독성이 좋은 것 같더라구요

**참고 URL**
https://velog.io/@dongdong98/JavaScript-%EC%83%81%EB%8C%80-%EA%B2%BD%EB%A1%9C%EB%A5%BC-%EA%B9%94%EB%81%94%ED%95%98%EA%B2%8C-%EA%B4%80%EB%A6%AC%ED%95%98%EB%8A%94-%EB%B2%95-%EB%AA%A8%EB%93%88-index.js-%ED%99%9C%EC%9A%A9

**Reply by yoopark** [reply]
@rockbell89 저도 깔끔해서 export * from 을 종종 썼었는데, 불필요한 파일을 모두 한번에 가져온다는 단점이 있더라구요.. 그 이후로 꼭 필요한 경우 아니라면 안 쓰긴 합니다. 

https://yoopark.github.io/export-from-trap
